python/tests/test_ccore.py

```diff
@@ -685,6 +686,42 @@ def test_Ensemble(Ensemble):
     assert es.member[1]['long'] == 7
     es.update_metadata(Metadata({'k': 'v'}))
     assert es['k'] == 'v'
+    # From here on we test features not in CoreEnsemble but only in
+    # LoggingEnsemble.   Note that we use pybind11 aliasing to
+    # define TimeSeriesEnsemble == LoggingEnsemble<TimeSeries> and
+    # SeismogramEnsemble == LoggingEnsemble<Seismogram>.
+    # Should be initially marked live
+    assert es.live()
+    es.elog.log_error("test_ensemble","test complaint",ErrorSeverity.Complaint)
+    es.elog.log_error("test_ensemble","test invalid",ErrorSeverity.Invalid)
+    assert es.elog.size() == 2
+    assert es.live()
+    es.kill()
+    assert es.dead()
+    # resurrect es
+    es.set_live()
+    assert es.live()
+    # validate checks for for any live members - this tests that feature
+    assert es.validate()
+    # need this temporary copy for the next test_
+    escopy=deepcopy(es)
```

Comment:
Well, there is nothing to sorry for. I just checked the error log, it seems the only error is at this line where it complains `deepcopy` is not defined. I guess it is not imported above? Also, if you are testing the serialization here, why not call pickle instead?

Reply:
It was MUCH worse than that.   Just pushed a complete fix with some bare bones tests in the test_ccore.py file.   I am not at all thrilled with this implementation, but to again quote that basic programming axiom "make it work before you make it fast".   In this case I'm very worried about both speed and memory use.   The deserialization section is a horrible memory pig and will create potential problems in some uses.   I think we ought to review this code in our conference call later today. 

The checks are still running as I'm writing this, but it passed my tests with pytest.   I have pytest only partly working, however, as I think spark is broken on this machine.